<a href="https://colab.research.google.com/github/vnguyen2011/VietAI-NTI_ChatGPTStreamlit/blob/main/ChatPDF_function.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 1 - CHAT PDF
Bài tập này tổng bao gồm 100 điểm, với 2 sections:
- Section 1 bao gồm TODO 1: 20 điểm tổng
- Section 2 bao gồm TODO 2, 3, 4 & 5: 80 điểm tổng

This exercise has 100 points total, consists of two sections:
- Section 1 includes TODO 1: Total of 20 points.
- Section 2 encompasses TODO 2, 3, 4, and 5: Total of 80 points.

## 1. Cài đặt và import các thư viện cần thiết

In [1]:
!pip install python-dotenv PyPDF2 streamlit langchain openai faiss-cpu

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 57.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 55.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.0/77.0 kB 8.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 54.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.6/190.6 kB 20.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.8/4.8 MB 64.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.1/82.1 kB 11.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.0/42.0 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 5.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 7.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This beha

In [2]:
from dotenv import load_dotenv
import os
from PyPDF2 import PdfReader
import streamlit as st
from langchain.text_splitter import CharacterTextSplitter
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain import FAISS
from langchain.chains.question_answering import load_qa_chain
from langchain.llms import OpenAI

## 2. Cấu hình OpenAI API Key
Sau khi đăng ký API key, các bạn chép key giống định dạng sau `sk-L7qQ...DN98XT` vào các chuỗi đang để trống bên dưới.

In [ ]:
import openai
openai.api_key = "" ## To configure OpenAI API
os.environ["OPENAI_API_KEY"] = "" ## To configure langchain connections with OpenAI

## 3. Chat với đoạn văn ngắn (Section 1)
Tiếp theo, chúng ta sẽ thử nghiệm việc chat với đoạn văn ngắn được lấy từ 1 trang PDF (short text/1 page) và hỏi bất cứ thứ gì về nội dung của đoạn văn này.

In [ ]:
# Specify the PDF file to be processed
pdf = "Deep Learning.pdf"

# Initialize an empty list to store chat history
chat_history = []

# Initialize a PdfReader object to read the PDF file
pdf_reader = PdfReader(pdf)

# Initialize an empty string to store the extracted text from the PDF
text = ""

# Iterate through the pages of the PDF and extract text from the first page only
for i, page in enumerate(pdf_reader.pages):
    text += f"### Page {i}:\n\n" + page.extract_text()
    break  # For a single page, we only take the first page

# Append a system message to the chat history to set the context
chat_history.append({"role": "system", "content": f"""You are an information retrieval assistant.
Text information: {text}

### Task: Answer questions using solely the information from the above text"""})

Tiếp theo, chúng ta sẽ định nghĩa hàm `ask_question` với đầu vào là câu hỏi `question`. Ngoài ra, hàm này còn khai thác lịch sử trò chuyện qua biến toàn cục `chat_history`.

In [ ]:
# Define a function to ask questions and interact with the AI model
def ask_question(question):
    global chat_history
    chat_history.append({"role": "user", "content": question})

    # Define parameters for the OpenAI ChatCompletion API
    params = dict(model="gpt-3.5-turbo", messages=chat_history)

    # Generate a response from the AI model
    response = openai.ChatCompletion.create(**params)
    content = response.choices[0].message.content
    chat_history.append({"role": "assistant", "content": content})

    # Print and return the assistant's response
    print(content)

Bắt đầu thử nghiệm thôi!

In [ ]:
ask_question("What's NN? Answer in less than 30 words")

NN stands for Neural Network, which is a computational model inspired by the structure and function of the human brain, used in deep learning for various applications.


In [ ]:
ask_question("Translate the above to Vietnamese")

NN viết tắt của Neural Network, đây là một mô hình tính toán được lấy cảm hứng từ cấu trúc và chức năng của não người, được sử dụng trong deep learning cho các ứng dụng khác nhau.


**[ENGLISH BELOW]**

**SECTION 1 (20 points): [TODO 1]**

**TODO 1: (20 points)**

- Hiện tại, hàm `ask_question` nối tiếp câu hỏi `question` và nội dung trả lời `content` vào biến toàn cục `chat_history`. Điều này sẽ giúp cho đoạn hội thoại có tính liên tục. Như ví dụ dịch nội dung trên, OpenAI chỉ dịch câu trả lời trước đó (về mạng NN) chứ không phải dịch tài liệu gốc. Tuy nhiên, hiện tại nó thiếu khả năng chỉnh sửa các cuộc trò chuyện trong quá khứ hoặc đặt lại cuộc trò chuyện. Cần phát triển tiếp các chức năng reset cuộc trò chuyện, delete hoặc edit các câu hỏi trước đó, tương tự như giao diện ChatGPT.
 - `ask_question("Translate the above to Vietnamese", reset_chat=True)` sẽ reset lịch sử chat
 - `ask_question("Translate the above to German", edit_last_question=True)` sẽ chỉnh sửa tại chỗ câu hỏi cuối cùng trong `chat_history` và lấy câu trả lời mới

------------
**[VIETNAMESE ABOVE]**

**SECTION 1 (20 points): [TODO 1]**

**TODO 1: (20 points)**

- Currently, the `ask_question` function appends the question `question` and the corresponding answer content `content` to the global variable `chat_history`. This ensures a continuous conversation flow, as demonstrated in the previous content translation example. OpenAI translates only the preceding answer (related to neural networks) rather than translating the original document. However, it currently lacks the capability to edit past conversations or reset the conversation. Further development is needed to introduce functions to reset the conversation, delete, or edit previous questions, similar to the ChatGPT interface.
   - `ask_question("Translate the above to Vietnamese", reset_chat=True)` will reset the chat history.
   - `ask_question("Translate the above to German", edit_last_question=True)` will allow editing of the last question in the `chat_history` and retrieving a new response.


# 4. Chat với đoạn văn dài (Section 2)
Do độ dài ngữ cảnh hạn chế, ta không thể đưa toàn bộ 100 trang vào các tương tác OpenAI của mình, cả về tính thực tế lẫn chi phí. Do đó, ta phải chia nhỏ các tài liệu dựa trên các yếu tố như kích thước văn bản hoặc các tiêu chí liên quan khác.

Trong trường hợp này, ta chia theo trang, mặc dù một số trang có số lượng token vượt quá giới hạn ký tự tiêu chuẩn là 4K token. Lựa chọn này cho phép ta tham chiếu chính xác trang nào đang được tương tác và thảo luận.

In [ ]:
# Specify the PDF file to be processed
pdf = "Deep Learning.pdf"

# Initialize a PdfReader object to read the PDF file
pdf_reader = PdfReader(pdf)

# Initialize an empty string to store the extracted text from the PDF
chunks = []

# Iterate through the pages of the PDF and extract text from each page
for i, page in enumerate(pdf_reader.pages):
    chunks.append(f"### Page {i}:\n\n" + page.extract_text())

# Print the number of pages in the PDF
print("Number of pages:", len(pdf_reader.pages))

Number of pages: 100


Tiếp theo, ta sẽ viết hàm `process_text` để xử lý toàn bộ nội dung các trang trong file PDF.

Chúng ta sẽ biểu diễn các trang sách bằng các vector ngôn ngữ (`embeddings = OpenAIEmbeddings()`). Khi trả lời câu hỏi, chúng ta đầu tiên sẽ chọn ra trang có embeddings gần nhất với câu hỏi `knowledgeBase.similarity_search`, và sẽ sử dụng thông tin trong trang tìm được để trả lời.

In [ ]:
!pip install tiktoken

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 8.7 MB/s eta 0:00:00


In [ ]:
# Define a function to process the extracted text
def process_text(chunks):
    # Convert the chunks of text into embeddings to form a knowledge base
    embeddings = OpenAIEmbeddings()
    knowledgeBase = FAISS.from_texts(chunks, embeddings)

    return knowledgeBase

# Process the extracted text to create a knowledge base
knowledgeBase = process_text(chunks)

In [ ]:
knowledgeBase

In [ ]:
# Define a question for which you want to find an answer
question = "What is Logistic regression cost function?"

# Use embeddings similarity search to search for the closest document (RAG)
docs = knowledgeBase.similarity_search(question)
llm = OpenAI()
chain = load_qa_chain(llm, chain_type='stuff')

# Run the question-answering chain to find the answer to the question
response = chain.run(input_documents=docs, question=question)

# Print the response
print(response)

 The Logistic regression cost function is J(w,b) = (1/m) * Sum(L(y'[i],y[i])) where L(y',y) = - (y*log(y') + (1-y)*log(1-y')).


In [ ]:
docs

[Document(page_content="Page 2:\n\nNeural Netw orks Basics\nLearn to set up a machine learning problem with a neural network mindset. Learn to use vectorization to speed up your\nmodels.\nBinar y classification\nMainly he is talking about how to do a logistic regression to make a binary classifier.\nImage taken from 3.bp.blogspot.com\nHe talked about an example of knowing if the current image contains a cat or not.\nHere are some notations:\nM is the number of training vectors\nNx is the size of the input vector\nNy is the size of the output vector\nX(1) is the first input vector\nY(1) is the first output vector\nX = [x(1) x(2).. x(M)]\nY = (y(1) y(2).. y(M))\nWe will use python in this course.\nIn NumPy we can make matrices and make operations on them in a fast and reliable time.\nLogistic r egression\nAlgorithm is used for classification algorithm of 2 classes.\nEquations:\nSimple equation: y = wx + b\nIf x is a vector: y = w(transpose)x + b\nIf we need y to be in between 0 and 1 (pr

**[ENGLISH BELOW]**

**SECTION 2 (80 points tổng): [TODO 2, 3, 4, & 5]**

Để phát triển các tính năng trong tương lai, ta đã vạch ra một số nhiệm vụ chính dự định thực hiện:

**TODO 2: (20 points)**
Hiện tại, mỗi khi người dùng đặt câu hỏi, ta sẽ thực hiện quy trình
`knowBase.similarity_search` để xác định trang có liên quan chặt chẽ nhất với câu hỏi xét về mức độ tương đồng nội dung. Điều này là cần thiết vì ta không thể chứa tất cả 100 trang trong giới hạn của OpenAI. Để nâng cao trải nghiệm người dùng, ứng dụng cho phép người dùng xác thực và tiếp tục cuộc trò chuyện của họ trong bối cảnh của cùng một trang, loại bỏ nhu cầu truy xuất tài liệu mới.

   - Input: `question`
   - Output: `def answer(question, retrieve_new_page=False)` - Cho phép người dùng tìm hiểu về trang hiện tại mà không cần chuyển đổi.

**TODO 3 (20 points): Chỉ chọn MỘT trong các ý 3.1 & 3.2**

**TODO 3.1:**
Hãy tưởng tượng một tình huống trong đó trang chính xác được truy xuất nhưng thông tin cần thiết cũng có thể được tìm thấy trên các trang tiếp theo. Để tạo điều kiện thuận lợi cho việc trả lời câu hỏi, ta sẽ triển khai một chức năng cho phép người dùng truy cập các trang tiếp theo để có thêm thông tin.

   - Input: `question`
   - Output: `def answer(question, retrieve_next_page=True)` - Cho phép người dùng đặt câu hỏi liên quan đến trang tiếp theo.

**TODO 3.2:**
Ta sẽ cung cấp cho người dùng khả năng linh hoạt để mở rộng các trang tài liệu khác nếu họ cho rằng câu trả lời hiện tại không chính xác hoặc nếu họ muốn tìm kiếm trong một trang cụ thể mà họ cho rằng có chứa thông tin liên quan.

   - Input: `question`
   - Output: `def answer(question, retrieve_page_id=[10, 11])` - Cho phép người dùng chọn lựa đến các trang cụ thể (vd., trang 10 hoặc 11) HOẶC `def answer(question, retrieve_diff_page=True)` để khám phá các trang khác nhau.

**TODO 4 (10 points):**

Ta lường trước đến việc cung cấp cho người dùng khả năng yêu cầu thông tin từ các nguồn bên ngoài, vượt ra ngoài giới hạn của các tệp PDF hiện tại.

   - Input: `question`
   - Output: `def answer(question, retrieve_external_knowledge=True)` - Cho phép người dùng tìm kiếm thông tin ngoài phạm vi của các file PDF hiện tại.

**TODO 5: (30 points)**

Để đảm bảo tính thân thiện với người dùng, chúng ta sẽ tạo ra một giao diện trực quan và tương tác cho ứng dụng này. Ta có thể tận dụng các công cụ như Streamlit hoặc Gradio để tạo ra ứng dụng tương tác hấp dẫn với người dùng.

   - Input: PDFs
   - Output: Giao diện người dùng cho phép người dùng tải lên các file PDF, xem nội dung và đặt câu hỏi một cách tương tác. Ta sẽ quay video giới thiệu ứng dụng. Đánh giá sẽ dựa trên tính thân thiện với người dùng, khả năng trình diễn và tính tương tác. Những bài nộp đặc biệt có thể được trình bày trong ngày giới thiệu lớp học của mình và được giới thiệu trên trang Facebook VietAI.

Những sáng kiến phát triển này nhằm mục đích nâng cao chức năng, tính linh hoạt và trải nghiệm người dùng.

----------------------------------
**[VIETNAMESE ABOVE]**

**SECTION 2 (80 points total): [TODO 2, 3, 4, & 5]**

In terms of our future development goals, we have outlined several key tasks that we plan to undertake:

**TODO 2: (20 points)**

Currently, each time a user poses a question, we execute the `knowledgeBase.similarity_search` process to identify the page most closely related to the question in terms of embeddings similarity. This is necessary because we cannot accommodate all 100 pages within OpenAI's constraints. To enhance the user experience, we aim to introduce functionality that allows users to validate and continue their chat within the context of the same page, eliminating the need to retrieve a new document.


**TODO 3 (20 points): Choose ONLY 1 option between 3.1 & 3.2**

**TODO 3.1:**

Imagine a scenario where the correct page has already been retrieved, but the required information might also be found on subsequent pages. To facilitate this exploration, we plan to implement a function that permits users to access the next pages for additional information.

   - Input: `question`
   - Output: `def answer(question, retrieve_next_page=True)` - Allows users to ask questions related to the next pages.

**TODO 3.2:**

We will provide users with the flexibility to explore entirely new pages if they believe the current answer is incorrect or if they wish to search in a specific page they believe contains relevant information.

   - Input: `question`
   - Output: `def answer(question, retrieve_page_id=[10, 11])` - Allows users to select specific pages (e.g., page 10 or 11) OR `def answer(question, retrieve_diff_page=True)` to explore different pages.

**TODO 4 (10 points):**

Our vision includes offering users the ability to request information from external sources, extending beyond the confines of the current PDFs.

   - Input: `question`
   - Output: `def answer(question, retrieve_external_knowledge=True)` - Allow users to search for information beyond the scope of the current PDF files.

**TODO 5: (30 points)**

As a demonstration of our commitment to user-friendliness, we aspire to create an intuitive and interactive interface for this application. We are considering leveraging tools like Streamlit or Gradio to achieve a seamless and engaging user interaction.

   - Input: PDFs
   - Output: A user interface enabling users to upload PDF files, view the content, and interactively ask questions. We will record a video showcasing the application. Evaluation will be based on user-friendliness, demonstrability, and interactivity. Exceptional submissions may be presented on our class demo day and featured on the VietAI Facebook page.

These development initiatives are aimed at enhancing the functionality, versatility, and user experience of our application. We are committed to continually improving the system to meet the evolving needs of our users.